liblwgeom/lwgeom_transform.c

```diff
@@ -172,3 +205,238 @@ lwproj_from_string(const char *str1)
+				return LW_FAILURE;
+			break;
+		}
+		case POLYGONTYPE:
+		{
+			LWPOLY *g = (LWPOLY*)geom;
+			for (i = 0; i < g->nrings; i++)
+			{
+				if (!ptarray_transform(g->rings[i], pj))
+					return LW_FAILURE;
+			}
+			break;
+		}
+		case MULTIPOINTTYPE:
+		case MULTILINETYPE:
+		case MULTIPOLYGONTYPE:
+		case COLLECTIONTYPE:
+		case COMPOUNDTYPE:
+		case CURVEPOLYTYPE:
+		case MULTICURVETYPE:
+		case MULTISURFACETYPE:
+		case POLYHEDRALSURFACETYPE:
+		case TINTYPE:
+		{
+			LWCOLLECTION *g = (LWCOLLECTION*)geom;
+			for (i = 0; i < g->ngeoms; i++)
+			{
+				if (!lwgeom_transform(g->geoms[i], pj))
+					return LW_FAILURE;
+			}
+			break;
+		}
+		default:
+		{
+			lwerror("lwgeom_transform: Cannot handle type '%s'",
+			          lwtype_name(geom->type));
+			return LW_FAILURE;
+		}
+	}
+	return LW_SUCCESS;
+}
+
+static int
+proj_crs_is_swapped(const PJ* pj_crs)
+{
+	PJ *pj_cs = proj_crs_get_coordinate_system(NULL, pj_crs);
+	int axis_count = proj_cs_get_axis_count(NULL, pj_cs);
+	if (axis_count > 0)
+	{
+		const char *out_name, *out_abbrev, *out_direction;
+		double out_unit_conv_factor;
+		const char *out_unit_name, *out_unit_auth_name, *out_unit_code;
+		/* Read only first axis, see if it is degrees / north */
+		proj_cs_get_axis_info(NULL, pj_cs, 0,
+			&out_name,
+			&out_abbrev,
+			&out_direction,
+			&out_unit_conv_factor,
+			&out_unit_name,
+			&out_unit_auth_name,
+			&out_unit_code
+			);
+		return (strcasecmp(out_direction, "north") == 0);
+	}
+	return LW_FALSE;
+}
+
+
+int
+ptarray_transform(POINTARRAY* pa, PJ* pj)
+{
+	uint32_t i;
+	POINT4D p;
+	size_t n_converted;
+	size_t n_points = pa->npoints;
+	size_t point_size = ptarray_point_size(pa);
+	int has_z = ptarray_has_z(pa);
+	double *pa_double = (double*)(pa->serialized_pointlist);
+
+	/* XXXX TODO check that the PJ has decimal degrees as units in input/output */
+
+	int input_swapped = proj_crs_is_swapped(proj_get_source_crs(NULL, pj));
+	int output_swapped = proj_crs_is_swapped(proj_get_target_crs(NULL, pj));
+
+	/* Convert to radians if necessary */
+	if (proj_angular_input(pj, PJ_FWD))
+	{
+		for (i = 0; i < pa->npoints; i++)
+		{
+			getPoint4d_p(pa, i, &p);
+			to_rad(&p);
+		}
+	}
+
+	if (input_swapped)
+		ptarray_swap_ordinates(pa, LWORD_X, LWORD_Y);
+
+	// size_t proj_trans_generic(PJ *P, PJ_DIRECTION direction,
+	// double *x, size_t sx, size_t nx,
+	// double *y, size_t sy, size_t ny,
+	// double *z, size_t sz, size_t nz,
+	// double *t, size_t st, size_t nt)
+
+	n_converted = proj_trans_generic(
+		pj, PJ_FWD,
+		pa_double,     point_size, n_points, /* X */
+		pa_double + 1, point_size, n_points, /* Y */
+		has_z ? pa_double + 2 : NULL,
+		has_z ? point_size : 0,
+		has_z ? n_points : 0, /* Z */
+		NULL, 0, 0 /* M */
+		);
+
+	if (n_converted != n_points)
+	{
+		lwerror("ptarray_transform: converted (%d) != input (%d)",
+			n_converted, n_points);
+		return LW_FAILURE;
+	}
+
+	int pj_errno_val = proj_errno(pj);
+	if (pj_errno_val)
+	{
+		lwerror("transform: %s (%d)",
```

Comment:
Should `transform` be `ptarray_transform` as in the error above? In any case, both could be changed by `%s` + `__func__`

Reply:
again I'm trying to match up messages across different proj backends... I can just remove the prefix entirely...